core/src/main/java/com/linecorp/armeria/client/circuitbreaker/AbstractCircuitBreakerClient.java

```diff
@@ -42,113 +37,68 @@
+        @Override
+        public String toString() {
+            return MoreObjects.toStringHelper(this)
+                              .add("factory", factory)
+                              .add("mapping", mapping)
+                              .toString();
+        }
     }
 
-    /**
-     * Creates a new instance that decorates the specified {@link Client}.
-     */
-    AbstractCircuitBreakerClient(Client<I, O> delegate, CircuitBreakerMapping mapping,
-                                 CircuitBreakerRuleWithContent<O> ruleWithContent) {
-        this(delegate, mapping, null, requireNonNull(ruleWithContent, "ruleWithContent"));
-    }
+    private final HandlerGenerator<?> handlerGenerator;
 
     /**
      * Creates a new instance that decorates the specified {@link Client}.
      */
-    private AbstractCircuitBreakerClient(Client<I, O> delegate, CircuitBreakerMapping mapping,
-                                         @Nullable CircuitBreakerRule rule,
-                                         @Nullable CircuitBreakerRuleWithContent<O> ruleWithContent) {
+    <CB> AbstractCircuitBreakerClient(Client<I, O> delegate, CircuitBreakerClientHandlerFactory<CB, I> factory,
+                                      ClientCircuitBreakerGenerator<CB> mapping) {
         super(delegate);
-        this.mapping = requireNonNull(mapping, "mapping");
-        this.rule = rule;
-        this.ruleWithContent = ruleWithContent;
-        if (ruleWithContent != null) {
-            fromRuleWithContent = fromCircuitBreakerRuleWithContent(ruleWithContent);
-        } else {
-            fromRuleWithContent = null;
-        }
-    }
-
-    /**
-     * Returns the {@link CircuitBreakerRule}.
-     *
-     * @throws IllegalStateException if the {@link CircuitBreakerRule} is not set
-     */
-    final CircuitBreakerRule rule() {
-        checkState(rule != null, "rule is not set.");
-        return rule;
-    }
-
-    /**
-     * Returns the {@link CircuitBreakerRuleWithContent}.
-     *
-     * @throws IllegalStateException if the {@link CircuitBreakerRuleWithContent} is not set
-     */
-    final CircuitBreakerRuleWithContent<O> ruleWithContent() {
-        checkState(ruleWithContent != null, "ruleWithContent is not set.");
-        return ruleWithContent;
-    }
-
-    /**
-     * Returns the {@link CircuitBreakerRule} derived from {@link #ruleWithContent()}.
-     *
-     * @throws IllegalStateException if the {@link CircuitBreakerRuleWithContent} is not set
-     */
-    final CircuitBreakerRule fromRuleWithContent() {
-        checkState(ruleWithContent != null, "ruleWithContent is not set.");
-        return fromRuleWithContent;
+        handlerGenerator = new HandlerGenerator<>(factory, mapping);
     }
 
     @Override
     public final O execute(ClientRequestContext ctx, I req) throws Exception {
-        final CircuitBreaker circuitBreaker;
+        final CircuitBreakerClientHandler<I> handler = handlerGenerator.acquire();
```

Comment:
`HandlerGenerator` is used only in one place here.
Wouldn't be simple to reference `CircuitBreakerClientHandlerFactory` and `ClientCircuitBreakerGenerator` directly in `AbstractCircuitBreakerClient`?
```suggestion
        final CircuitBreakerClientHandler<I> handler = factory.generateHandler(mapping);
```

Reply:
This was actually a result of me not wanting to define a generic type `CB` in `AbstractCircuitBreakerClient`. (To do this, we need to hold a reference to factory, mapping in `AbstractCircuitBreakerClient`.

I'm open to ideas to make this less complex though 😅

Reply:
> This was actually a result of me not wanting to define a generic type CB

Indeed. I want it too. I will delve into the relationship between the APIs. It seems possible somehow.